rest_api/auth.py

```diff
@@ -0,0 +1,112 @@
+import re
+from time import time
+import uuid
+from pydantic import BaseModel
+from quart import Blueprint, request, abort, current_app as app
+from quart_schema import validate_request
+from pydantic import Field
+from database import db, registration_blocked_ips
+import security
+
+auth_bp = Blueprint("auth_bp", __name__, url_prefix="/auth")
+time
+
+class AuthRequest(BaseModel):
+    username: str = Field(min_length=1, max_length=20)
+    password: str = Field(min_length=1, max_length=255)
+
+
+@auth_bp.post('/login')
+@validate_request(AuthRequest)
+async def login(data: AuthRequest):
+    for bucket_id in [
+        f"login:i:{request.ip}",
+        f"login:u:{data.username}:s",
+        f"login:u:{data.username}:f"
+    ]:
+        if security.ratelimited(bucket_id):
+            abort(429)
+
+    account = db.usersv0.find_one({"_id": data.username}, projection={
+        "tokens": 1,
+        "pswd": 1,
+        "flags": 1,
+        "permissions": 1,
+        "ban": 1,
+        "delete_after": 1
+    })
+
+    if not account:
+        abort(404)
+
+    if (account["flags"] & security.UserFlags.DELETED) or (account["delete_after"] is not None and account["delete_after"] <= time() + 60):
+        security.ratelimit(f"login:u:{data.username}:f", 5, 60)
+        return {"error": True, "type": "accountDeleted"}, 401
+
+    if (data.password not in account["tokens"]) and (not security.check_password_hash(data.password, account["pswd"])):
+        security.ratelimit(f"login:u:{data.username}:f", 5, 60)
+        abort(401)
+
+    db.netlog.update_one({"_id": {"ip": request.ip, "user": data.username}}, {"$set": {"last_used": int(time())}}, upsert=True)
+
+    security.ratelimit(f"login:u:{data.username}:s", 25, 300)
+
+    token = security.generate_token()
+
```

Comment:
there used to be something around here that sent an inbox message if an account deletion was cancelled, that should be re-implemented here